service/builder.go

```diff
@@ -22,27 +22,26 @@ package service
 
 import "github.com/pkg/errors"
 
-type moduleCreateInfo struct {
-	name             string
-	moduleCreateFunc ModuleCreateFunc
-	options          []ModuleOption
+type moduleInfo struct {
+	provider ModuleProvider
+	options  []ModuleOption
 }
 
 // A Builder is a helper to create a service
 type Builder struct {
-	options []Option
-	modules []*moduleCreateInfo
+	options     []Option
+	moduleInfos []*moduleInfo
 }
 
 // WithModule is a helper to create a service without any options
-func WithModule(name string, module ModuleCreateFunc, options ...ModuleOption) *Builder {
+func WithModule(provider ModuleProvider, options ...ModuleOption) *Builder {
```

Comment:
I was referring to [https://github.com/uber-go/fx/pull/353#discussion_r104001122](comment)